File: packages/cli/hieudoanm/go/src/libs/http/http.go
package http

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

const CONTENT_TYPE_HEADER = "Content-Type"
const CONTENT_TYPE_APPLICATION_JSON = "application/json"
const RESPONSE_ERROR = "Response Error"
const RESPONSE_STATUS = "Response Status"
const RESPONSE_BODY = "Response Body"

// Options
type Options struct {
	Header http.Header
	Query  map[string]string
	Body   map[string]interface{}
	Debug  bool
}

func debugLog(debug bool, label string, value interface{}) {
	if debug {
		fmt.Printf("%s: %+v\n", label, value)
	}
}

func setHeadersAndQuery(req *http.Request, options Options) {
	if options.Header != nil {
		req.Header = options.Header
	}
	if options.Query != nil {
		q := req.URL.Query()
		for k, v := range options.Query {
			q.Add(k, v)
		}
		req.URL.RawQuery = q.Encode()
	}
}

// Get ...
func Get(url string, options Options) ([]byte, error) {
	request, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}

	setHeadersAndQuery(request, options)

	client := &http.Client{}
	response, err := client.Do(request)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()

	body, err := io.ReadAll(response.Body)
	if err != nil {
		debugLog(options.Debug, RESPONSE_ERROR, err)
		return nil, err
	}

	debugLog(options.Debug, RESPONSE_STATUS, response.Status)
	debugLog(options.Debug, RESPONSE_BODY, string(body))
	return body, nil
}

// Post ...
func Post(url string, options Options) ([]byte, error) {
	requestData, err := json.Marshal(options.Body)
	if err != nil {
		return nil, err
	}

	request, err := http.NewRequest("POST", url, bytes.NewBuffer(requestData))
	if err != nil {
		return nil, err
	}

	setHeadersAndQuery(request, options)
	request.Header.Set(CONTENT_TYPE_HEADER, CONTENT_TYPE_APPLICATION_JSON)

	client := &http.Client{}
	response, err := client.Do(request)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()

	body, err := io.ReadAll(response.Body)
	if err != nil {
		debugLog(options.Debug, RESPONSE_ERROR, err)
		return nil, err
	}

	debugLog(options.Debug, RESPONSE_STATUS, response.Status)
	debugLog(options.Debug, RESPONSE_BODY, string(body))
	return body, nil
}

// Put ...
func Put(url string, options Options) ([]byte, error) {
	requestData, err := json.Marshal(options.Body)
	if err != nil {
		return nil, err
	}

	request, err := http.NewRequest("PUT", url, bytes.NewBuffer(requestData))
	if err != nil {
		return nil, err
	}

	setHeadersAndQuery(request, options)
	request.Header.Set(CONTENT_TYPE_HEADER, CONTENT_TYPE_APPLICATION_JSON)

	client := &http.Client{}
	response, err := client.Do(request)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()

	body, err := io.ReadAll(response.Body)
	if err != nil {
		debugLog(options.Debug, RESPONSE_ERROR, err)
		return nil, err
	}

	debugLog(options.Debug, RESPONSE_STATUS, response.Status)
	debugLog(options.Debug, RESPONSE_BODY, string(body))
	return body, nil
}

// Delete ...
func Delete(url string, options Options) ([]byte, error) {
	requestData, err := json.Marshal(options.Body)
	if err != nil {
		return nil, err
	}

	request, err := http.NewRequest("DELETE", url, bytes.NewBuffer(requestData))
	if err != nil {
		return nil, err
	}

	setHeadersAndQuery(request, options)
	request.Header.Set(CONTENT_TYPE_HEADER, CONTENT_TYPE_APPLICATION_JSON)

	client := &http.Client{}
	response, err := client.Do(request)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()

	body, err := io.ReadAll(response.Body)
	if err != nil {
		debugLog(options.Debug, RESPONSE_ERROR, err)
		return nil, err
	}

	debugLog(options.Debug, RESPONSE_STATUS, response.Status)
	debugLog(options.Debug, RESPONSE_BODY, string(body))
	return body, nil
}

File: packages/cli/hieudoanm/rust/src/libs/http.rs
use std::collections::HashMap;

pub const CONTENT_TYPE_HEADER: &str = "Content-Type";
pub const CONTENT_TYPE_APPLICATION_JSON: &str = "application/json";
pub const RESPONSE_ERROR: &str = "Response Error";
pub const RESPONSE_STATUS: &str = "Response Status";
pub const RESPONSE_BODY: &str = "Response Body";

#[derive(Debug, Default, Clone)]
pub struct Options {
    pub headers: HashMap<String, String>,
    pub query: HashMap<String, String>,
    pub body: Option<serde_json::Value>,
    pub debug: bool,
}

fn debug_log(debug: bool, label: &str, value: &str) {
    if debug {
        println!("{label}: {value}");
    }
}

fn set_headers_and_query(
    req: reqwest::RequestBuilder,
    options: &Options,
) -> reqwest::RequestBuilder {
    let mut req = req;
    for (k, v) in &options.headers {
        req = req.header(k, v);
    }
    req
}

pub async fn get(url: &str, options: Options) -> Result<Vec<u8>, anyhow::Error> {
    let client = reqwest::Client::new();
    let mut req = client.get(url);
    req = set_headers_and_query(req, &options);
    if !options.query.is_empty() {
        req = req.query(&options.query);
    }

    let response = req.send().await?;
    let body = response.bytes().await?.to_vec();

    debug_log(
        options.debug,
        RESPONSE_STATUS,
        &String::from_utf8_lossy(&body),
    );
    debug_log(
        options.debug,
        RESPONSE_BODY,
        &String::from_utf8_lossy(&body),
    );
    Ok(body)
}

pub async fn post(url: &str, options: Options) -> Result<Vec<u8>, anyhow::Error> {
    let client = reqwest::Client::new();
    let mut req = client.post(url);
    req = set_headers_and_query(req, &options);
    if !options.query.is_empty() {
        req = req.query(&options.query);
    }
    if let Some(body) = &options.body {
        req = req.json(body);
    }

    let response = req.send().await?;
    let body = response.bytes().await?.to_vec();

    debug_log(
        options.debug,
        RESPONSE_STATUS,
        &String::from_utf8_lossy(&body),
    );
    debug_log(
        options.debug,
        RESPONSE_BODY,
        &String::from_utf8_lossy(&body),
    );
    Ok(body)
}

pub async fn put(url: &str, options: Options) -> Result<Vec<u8>, anyhow::Error> {
    let client = reqwest::Client::new();
    let mut req = client.put(url);
    req = set_headers_and_query(req, &options);
    if !options.query.is_empty() {
        req = req.query(&options.query);
    }
    if let Some(body) = &options.body {
        req = req.json(body);
    }

    let response = req.send().await?;
    let body = response.bytes().await?.to_vec();

    debug_log(
        options.debug,
        RESPONSE_STATUS,
        &String::from_utf8_lossy(&body),
    );
    debug_log(
        options.debug,
        RESPONSE_BODY,
        &String::from_utf8_lossy(&body),
    );
    Ok(body)
}

pub async fn delete(url: &str, options: Options) -> Result<Vec<u8>, anyhow::Error> {
    let client = reqwest::Client::new();
    let mut req = client.delete(url);
    req = set_headers_and_query(req, &options);
    if !options.query.is_empty() {
        req = req.query(&options.query);
    }
    if let Some(body) = &options.body {
        req = req.json(body);
    }

    let response = req.send().await?;
    let body = response.bytes().await?.to_vec();

    debug_log(
        options.debug,
        RESPONSE_STATUS,
        &String::from_utf8_lossy(&body),
    );
    debug_log(
        options.debug,
        RESPONSE_BODY,
        &String::from_utf8_lossy(&body),
    );
    Ok(body)
}
